src/chain/chain.rs

```diff
@@ -371,9 +371,19 @@ impl Chain {
         // Check the lowest version of our info that any neighbour has and remove everything less
         // than it
         let our_oldest_ver = self.their_knowledge.values().min().map_or(0, |&v| v);
-        // TODO since the list is sorted according to the version, choose something more efficient
-        self.our_infos
-            .retain(|&(ref si, _)| *si.version() >= our_oldest_ver);
+        let oldest_ver_index = self
```

Comment:
First, the result from this calculation is not `the index of the oldest version` but shall be `the index of the first version older than our_oldest_ver` ? so the variable name is little bit mis-leading?

Second, in case we have a list of versions of [0, 1, 2, 3, 5, 8], and the our_oldest_ver to be 7.
The previous retain approach will give us a result of [8].
However the drain approach will give us [0, 1, 2, 3, 5, 8], because the result from binary_search is `Err(4)` which will then be replaced as `0` and cause no entry to be drained.
To make the behaviour consistent, maybe we need somthing like:
```
let oldest_ver_index = match self
    .our_infos
    .binary_search_by_key(&our_oldest_ver, |&(ref si, _)| *si.version())
{
    Ok(index) => index,
    Err(index) => cmp::min(index + 1, self.our_infos.len()),
};
```

If we don't want to do so, maybe worth to put a comment, mentioning the `non-equal` situation won't happen?

Reply:
I'll change this variable name to `version_retention_threshold_index` to be more clear.

In another discussion above, we agreed that it would be safer to not drop anything in the non-equal situation you described and add a `log_or_panic` to warn us if this pathological case arises. @madadam, @afck, thoughts?

Reply:
Yes, I agree with this change now.

Reply:
> First, the result from this calculation is not the index of the oldest version but shall be the index of the first version older than our_oldest_ver ? so the variable name is little bit mis-leading?

Done